schema.c

```diff
@@ -4065,3 +4367,553 @@ sch_json_to_gnode (sch_instance * instance, sch_node * schema, json_t * json, in
     }
     return root;
 }
+
+static gpointer
+copy_node_data (gconstpointer src, gpointer dummy)
+{
+    char *data = NULL;
+
+    if (src)
+        data = g_strdup (src);
+    return data;
+}
+
+static void
+sch_translate_data (sch_instance * instance, GNode *node, sch_xlat_data *xlat_data, int flags, bool incoming)
+{
+    int ret;
+    int s_0 = lua_gettop (instance->ls);
+
+    lua_getglobal (instance->ls, xlat_data->lua_module);
+    lua_getfield (instance->ls, -1, xlat_data->lua_func);
+    if (!lua_isnil (instance->ls, -1))
+    {
+        lua_pushstring (instance->ls, APTERYX_NAME (node));
+        lua_pushstring (instance->ls, APTERYX_NAME (node->children));
+        lua_pushstring (instance->ls, incoming ? "in" : "out");
+        ret = lua_pcall (instance->ls, 3, 1, 0);
+        if (ret)
+            sch_lua_error (instance->ls, ret);
+        else if (lua_gettop (instance->ls) != (s_0 + 2))
+        {
+            syslog (LOG_ERR, "XML LUA: Stack not zero(%d) after function: %s.%s\n",
+                    lua_gettop (instance->ls), xlat_data->lua_module,
+                    xlat_data->lua_func);
+            lua_pop (instance->ls, 1);
+        }
+        else
+        {
+            const char *value = lua_tostring (instance->ls, -1);
+            char *old_value = (char *) node->children->data;
+            node->children->data = g_strdup (value);
+            DEBUG (flags, "function: %s old value %s new value %s\n", xlat_data->lua_func, old_value, value);
+            g_free (old_value);
+        }
+    }
+    else
+    {
+        lua_pop(instance->ls, 1);
+        return;
+    }
+}
+
+static bool
+sch_translate_child_exists (GNode *parent, char *name, GNode **ret_node)
+{
+    GNode *child;
+
+    for (child = parent->children; child; child = child->next)
+    {
+        if (g_strcmp0 (APTERYX_NAME (child), name) == 0)
+        {
+            *ret_node = child;
+            return true;
+        }
+    }
+
+    *ret_node = NULL;
+    return false;
+}
+
+static bool
+sch_translate_sibling_exists (GNode *node, char *name, GNode **ret_node)
```

Comment:
use sch_node_child with node->parent

Reply:
sch_node_child is looking for a matching child in and xmlNode tree. sch_translate_sibling_exists is looking for a matching sigbling in a GNode tree.